src/Microsoft.AspNetCore.Blazor/Components/BlazorComponent.cs

```diff
@@ -169,14 +169,26 @@ private void HandleException(Exception ex)
             Console.Error.WriteLine($"[{ex.GetType().FullName}] {ex.Message}\n{ex.StackTrace}");
         }
 
-        void IHandleEvent.HandleEvent(UIEventHandler handler, UIEventArgs args)
+        Task IHandleEvent.HandleEvent(Func<UIEventArgs, Task> handler, UIEventArgs args)
         {
-            handler(args);
+            var task = handler(args);
 
             // After each event, we synchronously re-render (unless !ShouldRender())
             // This just saves the developer the trouble of putting "StateHasChanged();"
             // at the end of every event callback.
             StateHasChanged();
+
+            if (task.Status == TaskStatus.RanToCompletion)
+            {
+                return task;
+            }
+
+            return task.ContinueWith(OnAsyncTaskCompleted);
```

Comment:
Was this the flow that you had in mind? In the sample I updated things 'just work' without any calls to `StateHasChanged` in user code.

Reply:
Note that this calls `StateHasChanged` twice if you actually do async work.

Reply:
Yes, once synchronously when the task started, and once when the task has finished, right? That's the desired effect.

Reply:
Ok great, my memory was correct then 👍